[Woo POS] Coupons: Display discounted amount in price summary when coupon applied – Dummy UI

staskus opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #15336
Build on top of: https://github.com/woocommerce/woocommerce-ios/pull/15451
<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

<!-- 
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
🛠 Need an AI review?  
Add a comment: `@coderabbitai review`
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
-->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Adding additional discount totals line to totals breakdown. It's only shown if:
- Coupons are added to the cart AND totals are loading
OR
- Coupons are added to the cart AND discount is not zero

The totals will not fit now with extremely large accessibility sizes. I'm not addressing it until we finalize the UI.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

Test the cases:

- Add product and NO coupon -> Three loading lines are shown -> Loaded totals don't contain discount line
- Add product and a valid coupon -> Four loading lines are shown -> Loaded totals contain discount line
- Add product and a valid coupon that applies $0 discount -> Four loading lines are shown -> Loaded totals don't contain discount line

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

Additionally confirmed that the changes don't affect the view when the feature flag is disabled and coupons are not added.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/056f0971-e2c1-41db-854c-4afcf18305f9

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

staskus committed:
Add discountTotal to PointOfSaleOrderTotals

staskus committed:
Add shouldShowTotalDiscountField

We should show discounts field only if:

- Coupons in cart AND totals are loading
or
- Coupons in cart AND discount is not zero

staskus committed:
Display discount totals UI in TotalsView

staskus committed:
Update TotalsViewHelperTests.swift

iamgabrielma commented on WooCommerce/Classes/POS/ViewHelpers/TotalsViewHelper.swift:
```diff
@@ -73,4 +73,12 @@ final class TotalsViewHelper {
         }
     }
+
+    func shouldShowTotalDiscountField(cart: Cart, orderTotals: PointOfSaleOrderTotals?) -> Bool {
+        guard cart.coupons.isNotEmpty else {
+            return false
+        }
+
+        return orderTotals == nil || orderTotals?.discountTotal != nil
+    }
```
Should we add comments here regarding the logic? Or perhaps declare them explicitly? Eg:
```
let hasCoupons = cart.coupons.isNotEmpty
let hasDiscounts = orderTotals?.discountTotal != nil
... 
```

Question: The comment on e9f4ea56c58d307fb6d546c28b6fd4b239712fd2 shares that the discounts have to be non-zero, but this checks for non-nil discounts. Does this cover the same case? Or should be handled separately?

staskus replied:
> Question: The comment on https://github.com/woocommerce/woocommerce-ios/commit/e9f4ea56c58d307fb6d546c28b6fd4b239712fd2 shares that the discounts have to be non-zero, but this checks for non-nil discounts. Does this cover the same case? Or should be handled separately?

Yes, in this case, it's nil when the discount is zero. But it's not really clear. I'll try to organize and name it in a different.

iamgabrielma commented on WooCommerce/Classes/POS/Presentation/TotalsView.swift:
```diff
@@ -130,6 +130,15 @@ private extension TotalsView {
                               matchedGeometryId: Constants.matchedGeometrySubtotalId)
             Spacer().frame(height: Constants.subtotalsVerticalSpacing)
+
+            if viewHelper.shouldShowTotalDiscountField(cart: posModel.cart, orderTotals: orderTotals) {
```
Just wondering: 

We seem to create new instances of `TotalsViewHelper` and/or reference the existing one across the files indiscriminately, and that's fine as it's just stateless and basically functional: Should we make this object just a struct and call static functions? That should add more consistency of how we use it and clarify we don't need to retain a certain state for it to work.

staskus replied:
I agree. Our intention is to use these helpers as esentially View Models with immutable helper methods.

I think the best approach is to turn helpers into structs. If someone wants to make it stateful, they would need to declare mutating funcs and they would get compiler errors within the views with:

> Cannot use mutating member on immutable value: 'self' is immutable

iamgabrielma reviewed (approved):
Works nicely, added a couple of non-blocking comments :shipit:

staskus committed:
Turn POS view helpers into structs

Having struct enforces helper to be stateless and immutable when used within Views

staskus committed:
Improve readability of TotalsViewHelper

staskus committed:
Merge branch 'trunk' into task/15336-pos-coupons-display-total-discount
